docs/rules/assertions.md

```diff
@@ -162,23 +154,29 @@ rules:
       - description
 ```
 
-### `context` example
+### `if` example
 
 The following example asserts that PUT responses with HTTP status 200 or 201 cannot return an `application/pdf`content type.
-Without the `context`, the assertion would evaluate every MediaTypesMap including:
+Without the `if`, the assertion would evaluate every MediaTypeMap including:
 - Responses with all codes, including codes other than 200 or 201
 - Responses for all HTTP methods, including DELETE, GET, POST, and more.
-To restrict the evaluation, use the `context` feature to limit what will be evaluated.
+To restrict the evaluation, use the `if` feature to limit what will be evaluated.
 
 ```yaml
 assert/no-pdf-in-ok-response:
-  context:
-  - type: Operation
-    matchParentKeys: [put]
-  - type: Response
-    matchParentKeys: ['201', '200']
-  subject: MediaTypesMap
-  disallowed: ['application/pdf']
+  if:
+  - subject: Operation
+    filterByParentKeys: 
+      - put
+    defined: true
+  - subject: Response
+    filterByParentKeys: 
+      - '201'
+      - '200'
+    defined: true
+  subject: MediaTypeMap
```

Comment:
It is look a little bit confusing to have 3 subject.  Maybe it is better to rename subject that we specifically validation to something uniq - like `evaluateSubject` ?

Reply:
Agree, I was thinking - why can't we define simple "if any item with given filters exist" rule, which when met, generates proper error. For example, yesterday's meeting ([link](https://us02web.zoom.us/rec/play/Ov8F9Wt1zIrtNXJfu2fQ3kvWedY0XxY2Jm1_80XnyW-A8Y5qoA-lXkV7N0OYgF_w_99vbxLCNAugOYZs.LZHS2eQRSFkXVA-8?_x_zm_rhtaid=820&_x_zm_rtaid=FzJlBy6FQ5WUDELmSdfpiQ.1663920708235.aa1d24cc93317830c10d3cc84ae2ea9d&autoplay=true&continueMode=true&startTime=1663857448000)) __"media type not pdf"__ rule:

```
assert:
	name: media-type-map-not-pdf
	when:
		- subject: Operation
		  filterOutParentKeys:
			- put
		- subject: Response
		  filterInParentKeys: ['201', '200']
		- subject: MediaTypesMap
		  filterInChildKeys: ['application/pdf']
	then:
		error: "Media type should not be pdf"
```

Inside `when` we can define conditions and if they are met (to be precise - item meeting such conditions is met), we can proceed with `then` actions.

Reply:
What else can be in `then`?

Reply:
Can we have else?

Reply:
also, it's essential to understand that the `when` statements are not separate conditions, it defines the hierarchy:

```
Operation(put) -> Response(201, 200) -> MediaTypesMap
```

Reply:
For `then` I don't know, if nothing else, probably we could end up with

```
...
   then: "Media type should not be pdf"
```

or replace `then` with `error` (actually makes more sense to me).

I didn't think about any use-case for `else`, but if needed, we can add it later. What is more, in future we can add `or` and `and` operators:

```
assert:
	name: media-type-map-not-pdf
	when:
		-and:
			- subject: Operation
			  filterOutParentKeys:
				- put
			- subject: Response
			  filterInParentKeys: ['201', '200']
			- or:
				- subject: MediaTypesMap
				  filterInChildKeys: ['application/pdf']
				- subject: MediaTypesMap
				  filterInChildKeys: ['application/json']
	then:
		error: "Media type should not be pdf or json"
```


When it comes to hierarchy, maybe we should introduce `children` property? So for instance:

```
assert:
	name: media-type-map-not-pdf
	when:
		- subject: Operation
		  filterOutParentKeys:
			- put
		  children:
			- subject: Response
			  filterInParentKeys: ['201', '200']
			  children:
				- subject: MediaTypesMap
				  filterInChildKeys: ['application/pdf']
	then:
		error: "Media type should not be pdf"
```

Reply:
We can also think about introducing instead of `when` two operators: `whenAny` and `whenNone`. One would be for searching if there exists any item that matches given filter, another one would be to ensure that there is none item that matches condition.

Reply:
What about `whenAll`?

Reply:
Actually, `whenNone` is more or less `whenAll`, but naming is not the best, I believe it should be more like `whenAnyNot` - when there is any item that doesn't satisfy condition. For instance, if we want to ensure that all operations contain media type "application/json" we can write:

```
assert:
	name: 
	whenAnyNot:
		- subject: Operation
		  filterOutParentKeys:
			- put
		  children:
			- subject: Response
			  filterInParentKeys: ['201', '200']
			  children:
				- subject: MediaTypesMap
				  filterInChildKeys: ['application/json']
	error: "All operations should contain application/json media type"
```

Basically, all such operators can be added gradually, for MVP I believe `whenAny` would be sufficient.

Reply:
Thanks all for bringing your ideas 🙂 

Putting aside the exact naming, what do you think about the assertion structure? I mean `subject` + `condition` + `optional container for narrowing down which consists of subject + condition again`? Do some maybe have an idea of a better approach?

Reply:
I don't have a better idea.